### ISSUE TITLE
Set VTPR constants through the set_binary_interaction_parameter function

### ISSUE THREAD

**Issue by ianhbell** [CoolProp/CoolProp#1311]
This would be aij, aji... needed for fitting these parameters for VTPR models

**Comment by ibell** [comment]
This is a problem because the UNIFAQ library is a const reference. @JonWel do you have any ideas?

**Comment by JonWel** [comment]
The library is in deed a const.
However, `UNIFAQMixture` is not const, and it's where we need these parameters.
`set_interaction_parameters` is where the parameters (aij, aji, bij, bji, cij, cji) are extracted from the library.
So if we overload this `set_interaction_parameters` with a version where these parameters can be given as input, we can have what we want without needing to change the library.

We have to give the parameters with corresponding `mgi`, so it can place the coefficients in the right place in the `std::map<std::pair<int, int>, UNIFAQLibrary::InteractionParameters> interaction;`.

**Comment by JonWel** [comment]
I have made some changes in order to speed up VTPR in my branch [VTPR_speed](https://github.com/JonWel/CoolProp-f/tree/VTPR_speed).

Improvement is not as big as I wanted, but on a binary mixture and QT_INPUTS (I compute 200 points of the saturation curve), the time is reduced by 27%, which is already better (the before and after are both with -march=native), but still 10 times slower than PR for the same points.

One of the changes I made is to [set the interaction parameters in maps depending on the sgi pair](https://github.com/JonWel/CoolProp-f/blob/VTPR_speed/src/Backends/Cubics/UNIFAQ.cpp).
The `set_interaction_parameters`can be simplified as the first step already collect these interaction parameters. I did not touch it much right now as what comes up here may also change the way we want everything to be set up there, and that this part is not recursively executed.

**Comment by JonWel** [comment]
I have made some changes in order to speed up VTPR in my branch [VTPR_speed](https://github.com/JonWel/CoolProp-f/tree/VTPR_speed).

Improvement is not as big as I wanted, but on a binary mixture and QT_INPUTS (I compute 200 points of the saturation curve), the time is reduced by 27%, which is already better (the before and after are both with -march=native), but still 10 times slower than PR for the same computations.

One of the changes I made is to [set the interaction parameters in maps depending on the sgi pair](https://github.com/JonWel/CoolProp-f/blob/VTPR_speed/src/Backends/Cubics/UNIFAQ.cpp).
The `set_interaction_parameters`can be simplified as the first step already collect these interaction parameters. I did not touch it much right now as what comes up here may also change the way we want everything to be set up there, and that this part is not recursively executed.

**Comment by ibell** [comment]
I'm not sure if MSVC compiler is going to like the std::make_pair, but we shall see.  In any case, we can do that without the make_pair if MSVC doesn't like that.

A reduction of 27% is already quite good.  Sadly I don't think we can enable ``-march=native`` for the builds we make available to the public.  I don't think we will ever get very close to PR, but perhaps once we remove the numeric derivatives, and avoid the re-evaluation of terms we can get closer.

**Comment by ibell** [comment]
Now that i think about it, we need to provide ``mgi,mgi,aij,aji,.....``  That might make for an interesting API...

**Comment by JonWel** [comment]
MSVC 2015 do not seem to complain with `std::make_pair` (I made the changes from VS 2015).

With small other changes It's around 29% now.
`-march=native` do not seem to give much enhancement since some optimisations you made a while ago (ie your changes makes the normal binaries nearly as fast as with `-march=native` on my linux, no regressions):
3.27s vanilla vs 3.15s with native, so only 3-4% improvements to use `-march=native`on my test case. Nothing huge as before.

Current way iterates against the sgi, but the aij in deed only depends on the mgi. The current way of storage is thus not the shortest.
That's why I prefer that we see how to make the interface to store the aij before cleaning the code in `set_interaction_parameters`. It may be useful when we make this interface.

**Comment by ibell** [comment]
I'm not really concerned with MSVC2015 since that supports pretty much
completely C++11.  I'm more concerned with MSVC2009, which is needed to
compile python 2.7 wrapper

**Comment by JonWel** [comment]
I see.
I'm not yet a master on these specifications, it looked a quite convenient way to use, but if this not work (I don't have any 2009 to test with), I saw that you used an other way and I can change.
It may anyhow be good that I try to have a more coherent way trough the same file, so I may switch back to the other way to call.
Tell me if you have any preferences. I did not pull that one yet because I still want to work it a bit.

**Comment by ibell** [comment]
Well, if you push to another branch on github, you can make the tests run on that branch, and we can see if they cause failure of the builds on buildbot.  Alternatively, you can get the vc9 compiler for free for python 2.7, which is a stripped down version of msvc2009.

**Comment by JonWel** [comment]
The function we would use to set the interaction parameter is `set_binary_interaction_double`:
```C++
virtual void set_binary_interaction_double(const std::size_t i, const std::size_t j, const std::string &parameter, const double value)
```
For a given mixture, these parameters are linked to the main groups.
Thus, `i` and `j` should correspond to `mgi1` and `mgi2`, the main group indexes.

The respective values `aij`, `bij` and `cij` are stored in a map in the file `UNIFAQ.h`:
```C++
std::map<std::pair<int, int>, UNIFAQLibrary::InteractionParameters> interaction;
```
(May worth noting that the correct name of the method is UNIFAC "UNIQUAC Functional-group Activity Coefficients", this may be wise to correct it).

And the structure `InteractionParameters` is defined in `UNIFAQlibrary.h`:
```C++
    struct InteractionParameters{
        int mgi1, ///< The first main group index
            mgi2; ///< The second main group index
        double a_ij, ///< 
               a_ji, ///< 
               b_ij, ///< 
               b_ji, ///< 
               c_ij, ///< 
               c_ji; ///< 
        /// Swap a_ij with a_ji, b_ij with b_ji, etc.
        void swap() {
            std::swap(a_ij, a_ji);
            std::swap(b_ij, b_ji);
            std::swap(c_ij, c_ji);
        }
        /// Set all the values to 0
        void zero_out() {
            a_ij = 0; a_ji = 0;
            b_ij = 0; b_ji = 0;
            c_ij = 0; c_ji = 0;
        }
    };
```
So basically we need to add a function that allows to update the wanted element of the map:
```C++
if (parameter == "aij"){
    this->interaction[std::pair<int, int>(static_cast<int>(i), static_cast<int>(j))].a_ij  = value;
}
else if (parameter == "bij"){
    this->interaction[std::pair<int, int>(static_cast<int>(i), static_cast<int>(j))].b_ij  = value;
}
if (parameter == "cij"){
    this->interaction[std::pair<int, int>(static_cast<int>(i), static_cast<int>(j))].c_ij  = value;
}
```
I think that should work.
